### PR TITLE
feat: added Shift endpoint and Shift interface

### PR DESCRIPTION
**Pull request by Pandoyer**
issue #11,  I added placeholders for the 'Database' abstract class and the 'isUserAdmin' middleware function. The implementation was pretty much just to make the linter happy. 'shift-api' and 'shift-interface' are the real features.

### CODE REVIEW

File: test-app/backend/src/shift/shift-api.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import app from '../server';
+import isUserAdmin from '../middleware';
+import Database from '../models/database/database';
+import shift from './shift-interface';
+
+app.post('/api/new-shift', isUserAdmin, async (req, res) => {
+  console.log('POST: Creating Calendar Event...');
+
+  const { start, end, maxVolunteers, title, description, eventAdmin } = req.body as shift;
```
Comment (bglossner) [inline test-app/backend/src/shift/shift-api.ts:9]: The start and end dates here are actually going to be given as numbers, not `Date`s. We just need to convert those numbers to dates after confirming the type. Perhaps a function called like `shiftPreProcessor` that takes a `Partial<Shift>` and goes through and converts the numbers to actual dates and returns the edited shift.

File: test-app/backend/src/shift/shift-api.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import app from '../server';
+import isUserAdmin from '../middleware';
+import Database from '../models/database/database';
+import shift from './shift-interface';
+
+app.post('/api/new-shift', isUserAdmin, async (req, res) => {
+  console.log('POST: Creating Calendar Event...');
```
Comment (bglossner) [inline test-app/backend/src/shift/shift-api.ts:7]: Nit: Creating Shift

File: test-app/backend/src/shift/shift-interface.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+interface shift {
+  start: Date,
+  end: Date,
+  maxVolunteers: number,
+  title: string,
+  description: string,
```
Comment (bglossner) [inline test-app/backend/src/shift/shift-interface.ts:6]: If this is optional, we should put a `?` after it, like `description?: string`

File: test-app/backend/src/shift/shift-api.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
+import app from '../server';
```
Comment (bglossner) [inline test-app/backend/src/shift/shift-api.ts:1]: One thing I notice here is that while you are doing good error checking, there are still some assumptions. For the start/end dates for instance, you check if `start >= end`. But what if `start` or `end` is actually not even defined. We also need checks for that sort of thing, like with what you did for title. `if (!start || !end || start >= end) <some_error>`. There are other places that need this checked too.

**Review by bglossner — CHANGES_REQUESTED** [review]

**Comment by bglossner** [comment]
Also, sorry one more comment: make sure to write some unit tests!

**Comment by bglossner** [comment]
Glad to see a PR out for this. Good first iteration on the task!

**Comment by Pandoyer** [comment]
@bglossner got the tests to work with the express router. I have omitted the isUserAdmin middleware function because the tests are required to pass for a commit.

File: test-app/backend/test/models/shift/shift-api.test.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+const badVolShift = {
+  start: new Date().getTime(),
+  end: new Date().getTime(),
+  maxVolunteers: 0,
+  title: 'My Lovely Title',
+  description: 'I have a description',
+  eventAdmin: 'Adam Meza',
+};
+
+test('POST  /api/new-shift', async () => {
```
Comment (bglossner) [inline test-app/backend/test/models/shift/shift-api.test.ts:39]: Can you split each `await request(app)` into a new test?

The test description should be like `POST /api/new-shift; Shift fails to create due to bad no title provided`

File: test-app/backend/src/shift/shift-interface.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+  eventAdmin: string,
+  getShiftTime(): number
```
[the remark this answers is not in the document]
Comment (bglossner) [reply test-app/backend/src/shift/shift-interface.ts:8]: Checking in here again on what the `getShiftTime` does. I don't see it called anywhere.

**Review by bglossner — CHANGES_REQUESTED** [review]

File: test-app/backend/src/shift/shift-interface.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (Pandoyer) [reply test-app/backend/src/shift/shift-interface.ts]: It was part of the original issue but you're right it doesn't do anything right now

**Comment by Pandoyer** [comment]
@bglossner split up the tests and did without the getShiftTime() function

File: test-app/backend/test/models/shift/shift-api.test.ts [hunk 7]
```diff
@@ -36,31 +36,37 @@ const badVolShift = {
...
   await request(app)
     .post('/api/new-shift')
     .send(myShift)
     .expect(201)
     .then((response) => {
       expect(response.text).toBe('New shift successfully created');
     });
+});
 
+test('POST  /api/new-shift; Shift fails to create ecasue there is no title', async () => {
```
Comment (bglossner) [inline test-app/backend/test/models/shift/shift-api.test.ts:49]: Misspelling here.

File: test-app/backend/test/models/shift/shift-api.test.ts [hunk 8]
```diff
@@ -36,31 +36,37 @@ const badVolShift = {
...
   await request(app)
     .post('/api/new-shift')
     .send(badDateShift)
     .expect(400)
     .then((response) => {
       expect(response.text).toBe('Start/End dates are required and Start date must come before end date');
     });
+});
 
+test('POST  /api/new-shift; Shift fails becasuse max volunteers is 0', async () => {
```
Comment (bglossner) [inline test-app/backend/test/models/shift/shift-api.test.ts:69]: Misspelling here.

**Review by bglossner — CHANGES_REQUESTED** [review]

**Review by bglossner — APPROVED** [review]

**Pull request merged**